File: src/commands/policy_iter.c
#include <time.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <float.h>
#include "../lib/env.h"
#include "../lib/agent.h"
#include "../lib/json.h"
#include "policy_iter.h"

int calc_m_links(int *state, int max_links) {
    int result = 0;
    for (int i = 0; i < max_links; i++) {
        if (state[i] == 0) {
            break;
        }

        result += 1;
    }

    return result;
}

float calc_value(Environment *env, Agent *agent, int m_links, int *state, float action) {
    if (action == 0) {
        return DEFAULT_VALUE;
    }

    int *fail_state = transition_fail(m_links, state);
    int *succ_state = transition_succ(env, m_links, state, action);

    int m_links_fail = calc_m_links(fail_state, m_links);
    int m_links_succ = calc_m_links(succ_state, m_links + 1);


    float val_fail = get_value(agent, m_links_fail, fail_state);
    float val_succ = get_value(agent, m_links_succ, succ_state);

    return -1.0f + (1 - action) * val_fail + action * val_succ;
}

void update_bellman(Environment *env, Agent *agent, int m_links, int s_idx) {
    int *state = env->states[m_links][s_idx];
    float action = get_policy(agent, m_links, state);

    float value = calc_value(env, agent, m_links, state, action);
    set_value(agent, m_links, state, value);
}

void eval_policy(Environment *env, Agent *agent, float tol) {
    float delta = FLT_MAX;
    int its = 0;
    
    while (delta >= tol) {
        delta = -1;

        for (int m_links = env->n_links - 1; m_links >= 0; m_links--) {
            for (int i = 0; i < env->st_per_lev[m_links]; i++) {
                float old_val = get_value(agent, m_links, env->states[m_links][i]);
                update_bellman(env, agent, m_links, i);
                float new_val = get_value(agent, m_links, env->states[m_links][i]);

                delta = fmax(delta, fabs(old_val - new_val));
            }
        }
    }
}

void update_policy(Environment *env, Agent *agent, int m_links, int s_idx) {
    int *state = env->states[m_links][s_idx];
    float best_action = get_policy(agent, m_links, state);
    float best_value = get_value(agent, m_links, state);

    for (int i = 0; i < env->n_actions; i++) {
        float action = env->actions[i];
        if (get_ttl_env(env, action) < env->n_links - m_links) {
            continue;
        }

        float value = calc_value(env, agent, m_links, state, action);

        best_action = value > best_value ? action : best_action;
        best_value = value > best_value ? value : best_value;
    }

    set_policy(agent, m_links, state, best_action);
}

bool improve_policy(Environment *env, Agent *agent) {
    bool policy_stable = true;

    for (int m_links = env->n_links - 1; m_links >= 0; m_links--) {
        for (int i = 0; i < env->st_per_lev[m_links]; i++) {
            int *state = env->states[m_links][i];
            float old_action = get_policy(agent, m_links, state);
            update_policy(env, agent, m_links, i);
            float new_action = get_policy(agent, m_links, state);

            if (old_action != new_action) {
                policy_stable = false;
            }
        }
    }

    return policy_stable;
}

void output_policy(Environment *env, Agent *agent, const char *output_path) {
    int json_size = (env->n_links + 20) * agent->n_states;
    char *policy_json = (char*) malloc(json_size * sizeof(char));
    serialise_hashmap(agent->policy, policy_json);

    char *value_json = (char*) malloc(json_size * sizeof(char));
    serialise_hashmap(agent->value, value_json);

    HashMap *params = create_hashmap(PARAMS_SIZE);
    insert_data(params, "n_links", (float) env->n_links);
    insert_data(params, "n_actions", (float) env->n_actions);
    insert_data(params, "alpha", env->alpha);
    insert_data(params, "gamma", env->gamma);
    insert_data(params, "f_thresh", env->f_thresh);

    char *params_json = (char*) malloc(PARAMS_SIZE * 20 * sizeof(char));
    serialise_hashmap(params, params_json);

    char *output_json = (char*) malloc((2 * json_size + PARAMS_SIZE * 20) * sizeof(char));
    serialise_system(policy_json, value_json, params_json, output_json);

    if (output_path == NULL) {
        printf("%s\n", output_json);
    } else {
        adx_store_data(output_path, output_json);
    }

    free(value_json);
    free(policy_json);
    free(params_json);
}

void adx_store_data(const char *filepath, const char *data) {
    FILE *fp = fopen(filepath, "w");
    if (fp != NULL) {
        fputs(data, fp);
        fclose(fp);
    }
}

int policy_iter(const AdaptiveProtocolSystem *sys) {
    Environment *env = create_env(sys);
    int n_states = count_states(env);
    Agent *agent = create_agent(n_states, sys->n_actions);

    bool policy_stable = false;
    int iter = 0;
    clock_t start = clock();

    while (!policy_stable) {
        iter += 1;
        eval_policy(env, agent, sys->tol);
        policy_stable = improve_policy(env, agent);
    }

    clock_t end = clock();
    printf("Policy iteration complete within %d iterations!\n", iter);
    printf("Policy iteration took %.5f seconds!\n", ((float) end - start) / CLOCKS_PER_SEC);
    output_policy(env, agent, "./output.json");
    return 0;
}

File: OptimalNearTerm/src/adaptivealgo/cli/policy_iter.py
import argparse
import json
from typing import List
import numpy as np
import time
import wandb

from adaptivealgo.cli import cli_main
from adaptivealgo.lib.agent import Agent
from adaptivealgo.lib.env import Environment
from adaptivealgo.lib.state import State

def build_argument_parser():
    """
    Define the command-line arguments for running policy iteration
    """

    parser = argparse.ArgumentParser(
        description="""
        policy_iter: solves a Markov decision process to generate a required number of links
        in a near-term quantum network using an adaptive generation protocol.
        """
    )
    parser.add_argument(
        "--links",
        dest="n_links",
        type=int,
        default="2",
        help="Number of required links in the network.",
    )
    parser.add_argument(
        "--thresh",
        dest="f_thresh",
        type=float,
        default="0.3",
        help="Minimum fidelity threshold (F_thresh).",
    )
    parser.add_argument(
        "--actions",
        dest="actions",
        type=str,
        default="0.1, 0.7",
        help="Comma-separated list of protocol probabilities of success (e.g. 0.25, 0.5, 0.75).",
    )
    parser.add_argument(
        "--alpha",
        dest="alpha",
        type=float,
        default="1.0",
        help="Defines mapping between probability and fidelity F_i = 1 - alpha * p_i",
    )
    parser.add_argument(
        "--gamma",
        dest="gamma",
        type=float,
        default="0.1",
        help="Fixed parameter specifying exponential memory decay.",
    )
    parser.add_argument(
        "--tol",
        dest="tol",
        type=float,
        default="1e-20",
        help="Tolerance at which the policy is considered stable.",
    )
    parser.add_argument(
        "--output",
        dest="output_path",
        type=str,
        default=None,
        help="Path to save the policy.",
    )
    return parser

def eval_policy(env: Environment, agent: Agent, tol: float):
    """
    Policy evaluation step

    :param Environment env: The environment object
    :param Agent agent: The agent object
    :param float tol: The tolerance at which the policy is considered stable
    """

    delta = np.inf

    while delta >= tol:
        delta = -np.inf
        for s_idx in range(agent.n_states):
            if env.is_terminal(env.states[s_idx]):
                continue

            old_value = agent.get_value(env.states[s_idx])
            update_bellman(env, agent, s_idx)
            delta = max(delta, np.abs(agent.get_value(env.states[s_idx]) - old_value))

def update_policy(env: Environment, agent: Agent, s_idx: int):
    """
    Update the policy of a state

    :param Environment env: The environment object
    :param Agent agent: The agent object
    :param int s_idx: The index of the state in the list of states
    """

    state = env.states[s_idx]
    best_action = None
    best_value = -np.inf

    for a_idx in range(agent.n_actions):
        action = env.actions[a_idx]
        if env.get_ttl(action) < env.n_links - len(state):
            continue

        value = calc_value(env, agent, state, action)

        best_action = a_idx if value > best_value else best_action
        best_value = value if value > best_value else best_value

    new_policy = [0 for _ in range(agent.n_actions)]
    new_policy[best_action] = 1
    agent.set_policy(state, new_policy)

def improve_policy(env: Environment, agent: Agent) -> bool:
    """
    Policy improvement step

    :param Environment env: The environment object
    :param Agent agent: The agent object
    :returns bool: Whether the policy is stable
    """

    policy_stable = True

    for s_idx in range(agent.n_states):
        if env.is_terminal(env.states[s_idx]):
            continue

        old_policy = agent.get_policy(env.states[s_idx])
        update_policy(env, agent, s_idx)
        new_policy = agent.get_policy(env.states[s_idx])

        if not np.array_equal(old_policy, new_policy):
            policy_stable = False

    return policy_stable

def calc_value(env: Environment, agent: Agent, state: State, action: float) -> float:
    """
    Calculate the value of a state-action pair

    :param Environment env: The environment object
    :param State state: The state
    :param float action: The action
    """

    fail_state, succ_state = env.transition(state, action)
    return -1 + (1 - action) * agent.get_value(fail_state) + action * agent.get_value(succ_state)

def update_bellman(env: Environment, agent: Agent, s_idx: int):
    """
    Update the value of a policy using the Bellman equation

    :param Environment env: The environment object
    :param Agent agent: The agent object
    :param int s_idx: The index of the state in the list of states
    """

    state = env.states[s_idx]
    value = 0

    for a_idx in range(agent.n_actions):
        action = env.actions[a_idx]
        if env.get_ttl(action) < env.n_links - len(state):
            continue

        prob = agent.get_policy(state)[a_idx]
        value += prob * calc_value(env, agent, state, action)

    agent.set_value(state, value)

def print_policy(agent: Agent, env: Environment):
    """
    Print the policy of the agent

    :param Agent agent: The agent object
    :param Environment env: The environment object
    """

    print("Final policy:")
    for state in env.states:
        if env.is_terminal(state):
            continue

        print(f"State {state}: {agent.get_policy(state)}")

def build_policy_dict(agent: Agent, env: Environment) -> dict:
    """
    Build a dictionary representation of the policy

    :param Agent agent: The agent object
    :param Environment env: The environment object
    :returns dict: The policy dictionary
    """

    policy = {}
    for state in env.states:
        if env.is_terminal(state):
            continue

        policy[str(state)] = int(np.argmax(agent.get_policy(state)))
    
    return policy

def find_policy(n_links: int, f_thresh: float, ps: List[float], alpha: float, gamma: float, tol: float, to_print: bool=False) -> dict:
    env = Environment(n_links, ps, f_thresh, alpha, gamma)
    agent = Agent(n_states=len(env.states), n_actions=len(ps))

    policy_stable = False
    i = 0

    while not policy_stable:
        i += 1
        eval_policy(env, agent, tol)
        policy_stable = improve_policy(env, agent)

    policy = build_policy_dict(agent, env)
    result = {
        "n_links": env.n_links,
        "f_thresh": env.f_thresh,
        "actions": env.actions,
        "alpha": env.alpha,
        "gamma": env.gamma,
        "policy": policy,
    }

    if to_print:
        print("Final values:")
        for state in env.states:
            print(f"State {state}: E[t] = {-agent.get_value(state)}")
        print_policy(agent, env)

    return result, i

def run(n_links: int, f_thresh: float, actions: str, alpha: float, gamma: float, tol: float, output_path: str):
    """
    Perform policy iteration given the command-line arguments

    :param int n_links: Target required number of links in the network
    :param float f_thresh: Minimum fidelity threshold
    :param str actions: Comma-separated list of probabilities
    :param float alpha: Specifies relationship between probablity and fidelity
    :param float gamma: Exponential memory decay constant
    """

    ps = [float(p) for p in actions.replace(" ", "").split(",")]

    wandb.init(
        project="adaptive-algorithms",
        config={
            "n_links": n_links,
            "f_thresh": f_thresh,
            "actions": ps,
            "alpha": alpha,
            "gamma": gamma,
            "tol": tol,
        }
    )
    
    start_time = time.process_time()
    policy, i = find_policy(n_links, f_thresh, ps, alpha, gamma, tol, to_print=not output_path)
    end_time = time.process_time()

    wandb.log({"policy_iterations": i, "time": end_time - start_time})
    wandb.finish()

    print(f"Policy iteration converged after {i} steps for {end_time - start_time} seconds")
    if output_path:
        with open(output_path, "w") as file:
            json.dump(policy, file)
    else:
        print(json.dumps(policy, indent=4))

def main():
    cli_main(build_argument_parser, run)
